### PR TITLE
dasOPENAI: streamed tool-call reassembly + offline streaming test

### PR DESCRIPTION
**Pull request by borisbat**
Lands the streaming follow-up promised on the merged PR #2938 review thread: `chat_stream` now reassembles streamed tool/function calls, and the SSE path gets **deterministic offline coverage** via a local streaming mock fixture (previously only a live test exercised it).

## What's here

- **Streamed tool-call reassembly** in `chat_stream` — fragments arrive split by `index` (id/name in the first frame, JSON `arguments` accreted across later frames); they're merged per-index into a new `ChatStreamResult.tool_calls`. Adds the delta wire types `ToolCallDelta` / `ToolCallFunctionDelta` and a `tool_calls` field on `ChatMessageDelta`. Negative/malformed `index` fragments are dropped (`continue if (idx < 0)`). Docstring drops the old "not reassembled" caveat.
- **Local streaming mock fixture** `modules/dasOPENAI/tests/_openai_stream_mock.das` (next to the tests, not in the tutorials tree) — serves `/v1/chat/completions` only and selects an SSE script from the request body: a fragmented tool-call sequence (with a leading malformed `index = -1` frame) when `tools` is present, a CRLF + `: ping` comment-line stream when the model id contains `stream-crlf`, else plain content. Frames are built from the typed structs so mock and client agree on wire shape; the leading `_` keeps dastest from treating it as a test.
- **Offline end-to-end test** `modules/dasOPENAI/tests/test_chat_stream_mock.das` stands up the fixture on a worker thread and runs real `chat_stream` calls, asserting content accumulation, CRLF/comment tolerance, and tool-call reassembly (`get_weather({"location":"Paris"})`) — including that the malformed `index = -1` frame is dropped without crashing or polluting the real call. The live `test_chat_stream.das` stays as a smoke test.
- **Tutorial mock server** (`tutorial_openai_server.das`) gains a fragmented tool-call streaming branch (plain content + tool-call — the two modes the tutorial demos use). **Tutorial OPENAI-07** + its RST gain a streamed-tool-calls section.
- **CLAUDE.md** — documents `require` path resolution (bare same-dir vs file-relative with explicit `.das` extension), discovered wiring up the fixture.

Adding the `tool_calls` array makes the streamed-chunk structs non-copyable; construction sites move-init (`delta <- …`) and `chat_stream` returns by move accordingly.

## Validation

| Check | Result |
|---|---|
| Offline `test_chat_stream_mock` (incl. negative-index guard) | 3 tests / 7 assertions pass |
| Existing `test_chat_stream` (unit + live) | 6/6 |
| Tutorial 07 end-to-end (content + tool call) | pass |
| All 7 tutorials `-dry-run` | clean |
| Lint (changed `.das`) | 0 issues |
| das2rst reference | 0 stubs, 0 Uncategorized |
| Clean Sphinx build | build succeeded, 0 warnings |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: modules/dasOPENAI/tests/_openai_stream_mock.das [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+// client's trailing-\r stripping and comment-line skipping.
+def private mock_stream_body_crlf() : string {
+    let words = ["CRLF", " stream", " works."]
+    var frames : array<string>
+    frames |> reserve(length(words) + 3)
+    frames |> push(": ping\r\n\r\n")
+    for (w in words) {
+        frames |> push("data: {content_chunk(w)}\r\n\r\n")
+    }
+    frames |> push("data: {finish_chunk("stop")}\r\n\r\n")
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasOPENAI/tests/_openai_stream_mock.das:71]: Not an issue — nested double-quotes inside a string-interpolation `{...}` are valid in daslang. `"data: {finish_chunk("stop")}\r\n\r\n"` compiles cleanly (lint reports 0 issues) and the test passes: `test_chat_stream_mock` asserts `res.finish_reason == "stop"`, which is only set if that exact frame parsed and round-tripped correctly. So the interpolation is parsed as `finish_chunk("stop")` as intended. No change needed.

File: modules/dasOPENAI/openai/openai_chat.das [hunk 2]
```diff
@@ -196,6 +214,16 @@ def chat_stream(client : OpenAIClient; var req : ChatCompletionRequest;
                                     parts |> push(d)
                                     invoke(on_delta, d)
                                 }
+                                for (tcd in frame.choices[0].delta.tool_calls) {
+                                    let idx = tcd.index
+                                    while (length(tools) <= idx) {
+                                        tools |> emplace(ToolCall())
+                                    }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasOPENAI/openai/openai_chat.das:229]: Good catch — fixed. `chat_stream` now guards the streamed index before growing/indexing `tools`: `continue if (idx < 0)` drops malformed fragments. Added a regression: the mock's tool-call stream now leads with an `index = -1` frame (bogus id/name/args), and the test asserts the result still has exactly one tool call with the correct id/arguments — proving the guard skips the malformed fragment without crashing or polluting the real call at index 0. 6/6 still green.

File: tutorials/dasOPENAI/tutorial_openai_server.das [hunk 3]
```diff
@@ -138,6 +170,9 @@ class MockOpenAIServer : HvWebServer {
             let body = string(req.body)
             if (find(body, "\"stream\":true") >= 0) {
                 set_header(resp, "Content-Type", "text/event-stream")
+                if (find(body, "\"tools\":") >= 0) {
+                    return resp |> TEXT_PLAIN(mock_stream_tool_calls_body())  // fragmented tool-call stream
+                }
                 return resp |> TEXT_PLAIN(mock_stream_body())
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tutorials/dasOPENAI/tutorial_openai_server.das:176]: Correct — the code is right; the PR description was stale. The CRLF + comment-line variant was moved into the test fixture (`_openai_stream_mock.das`) when the test was decoupled from the tutorials tree, so the tutorial server now only needs plain content + tool-call streaming (the two modes its own tutorial demos use). I've updated the PR description to match: tutorial server = plain + tool-call; the CRLF/comment + negative-index edge cases live in the test fixture.

File: modules/dasOPENAI/openai/openai_chat.das [hunk 4]
```diff
@@ -196,6 +214,17 @@ def chat_stream(client : OpenAIClient; var req : ChatCompletionRequest;
                                     parts |> push(d)
                                     invoke(on_delta, d)
                                 }
+                                for (tcd in frame.choices[0].delta.tool_calls) {
+                                    let idx = tcd.index
+                                    continue if (idx < 0)  // ignore malformed fragments (guard array index)
+                                    while (length(tools) <= idx) {
+                                        tools |> emplace(ToolCall())
+                                    }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasOPENAI/openai/openai_chat.das:229]: Fixed — added an upper bound too. The guard is now `continue if (idx < 0 || idx >= MAX_STREAM_TOOL_CALLS)` (`MAX_STREAM_TOOL_CALLS = 256`, far above any legitimate tool-call count), so an absurd index can't drive the grow-loop into unbounded allocation. Extended the regression: the mock's tool-call stream now leads with both an `index = -1` and an `index = 9999999` frame; the test asserts the result still has exactly one clean tool call, proving both out-of-range fragments are dropped. 6/6 green.
